utilities/partitioning/partition_validator.py

```diff
@@ -0,0 +1,167 @@
+##
+# @file partition_validator.py
+#
+# @brief This program validates an input partition file for use in the ngen framework
+#
+# @section TODO
+# - Update remote sections to show full communication path source->nexus->destination
+# - Update validate of remote sections to use full paths when present
+#
+# @section authors
+# - Created 10/26/2021 donald.w.johnson@noaa.gov
+
+import json
+import argparse
+
+
+def test_catchments_ids(data: dict):
+    """This function check the catchment ids stored in each partition of data to ensure that catchments are all in only
+        one partition
+
+        @param data A dictionary containing a list of partition records stored with the key 'partitions'
+
+        """
+    cat_ids = {}
+
+    # create a set of each partitions catchment_ids
+    for partition in data['partitions']:
+        cat_ids[partition["id"]] = set(partition["cat-ids"])
+
+    cat_id_conflicts = 0
+
+    # make sure that each catchment has a unique set of catchment_ids
+    for key1, cachment_set1 in cat_ids.items():
+        for key2, catchment_set2 in cat_ids.items():
+            if key1 != key2:
+                inter = cachment_set1.intersection(catchment_set2)
+                if len(inter) > 0:
+                    cat_id_conflicts = cat_id_conflicts + 1
+                    print("partition " + str(key1) + " and partition " + str(key2) + " both contain" + str(inter))
+
+    return cat_id_conflicts
+
+
+def test_nexus_ids(data: dict):
+    """This function check the nexus ids in the recorded partitions to determine the set of remote nexi
+
+        @param data A dictionary containing a list of partition records stored with the key 'partitions'
+
+        """
+    nex_ids = {}
+
+    # create a set of each partitions catchment_ids
+    for partition in data['partitions']:
+        nex_ids[partition["id"]] = set(partition["nex-ids"])
+
+    remote_nexus_ids = set()
+
+    # find the remote nexi by finding all nexus ids that are more than one partition
+    for key1, cachment_set1 in nex_ids.items():
+        for key2, catchment_set2 in nex_ids.items():
+            if key1 != key2:
+                inter = cachment_set1.intersection(catchment_set2)
+                if len(inter) > 0:
+                    remote_nexus_ids = remote_nexus_ids.union(inter);
+
+    return remote_nexus_ids
+
+
+def test_remote_connections(data: dict):
+    """This function checks to ensure that there exist a minal set of reciving remote connection sections, such
+       that MPI send and recives will pair properly
+
+        @param data A dictionary containing a list of partition records stored with the key 'partitions'
+
+        TODO update logic when partition files encode full communication path in recmote sections
+
+        """
+    partitions = data['partitions']
+
+    mpi_rank = -1;
+    err_count = 0
+    for partition in partitions:
+        mpi_rank = mpi_rank + 1
```

Comment:
I'm not sure this logic is correct.  

This assumes partitions within the config will be ordered by `id`, but I don't think we can do this.  At minimum, we'd need to add validation of an ordering constraint like that by checking `id` against the incremented `mpi_rank` value.  

However, we may want to throw out this constraint and just replace `mpi_rank` with `id`.   I don't recall the specific rules off the top of my head, but I have seen that JSON formatting tools sometimes rearrange the ordering of properties and elements.  It would be frustrating if standard tools couldn't be used to help make a partitioning config more readable, because they might invalidate it.

Reply:
I'm assuming that the partitions being in order of `id` is not a constraint that needs to be validated (for the functioning of the software), but if not then yes it shouldn't be assumed...I'm also not sure why `id` and `mpi-rank` are synonymous but named differently, but probably not worth changing pervasively without some additional refactor reason. A bigger question to me is whether the `id`s/`mpi_ranks` must be _contiguous_ and start with zero...maybe _that_ should be validated if it is required.

Concur that the best answer here is probably just to use the `id` instead of incrementing an `mpi_rank` variable...if these can actually be different, then at a minimum ordering the `partitions` array should be done...but if they can be different I don't know what that means or how the mapping of the two values is really done.

Notably, in JSON the order of key-value pairs between `{}` definitely can be reordered, that order is undefined... however, the order of elements within `[]` should always be preserved by any processor, order matters here. So in this case I don't think that's a concern, the order of elements _should_ be preserved regardless of any pre-processing.

Reply:
When the catchment file is processed each mpi id will look for catchment section with id==mpi_rank. I would have to check the logic to see if there is an assumption that the catchment ids are sorted.

Reply:
> I'm also not sure why id and mpi-rank are synonymous but named differently, but probably not worth changing pervasively without some additional refactor reason

The primary reason is just to avoid inconsistencies.  Remote connections are defined by a remote rank.  But their validity depends on the details of the rank's partition, and we must assume which one that is.  I think we should eventually address this more thoroughly.   We may be able to put that off, but ... 

> A bigger question to me is whether the ids/mpi_ranks must be contiguous and start with zero...maybe that should be validated if it is required.

This is interesting.  Say we have an example config with 4 partitions that "skips" partition/rank 2 (i.e., it has 0, 1, 3, and 4).  We have to consider what criteria we are using for determining validity:

* Option 1 : whether this config works well in the framework
  * At least with how things work now, this would (at best) be inefficient in rank utilization, so consider it not valid
* Option 2 :  whether this config works at all in the framework  
  * I'm not sure but strongly suspect it would break things, so consider it probably not valid
* Option 3: whether the entries are internally consistent within the config
  * In this case, correct me if I'm wrong, but I'm pretty sure this can be valid

This may seem convoluted, but unless we are using Option 3 (I don't think we really want to but am open to discussing), then validity cannot be established without some runtime details.  

If we are assuming ranks start at zero and are contiguous, we really do need to check for that.  The validations routines would also need to know how many total ranks/processes are involved.  Alternatively, validations routines would need a full list of the expected ranks.